File: src/source.rs
pub struct Source<'a> {
    pub code: &'a str,
    pub meta: Meta<'a>,
}

impl<'a> Source<'a> {
    pub fn new(name: &'a str, code: &'a str) -> Self {
        Self {
            meta: meta(name, code),
            code,
        }
    }
}

pub struct Meta<'a> {
    pub name: &'a str,
    pub lines: Vec<&'a str>,
}

#[derive(Clone, Copy)]
pub struct Location<'a> {
    pub start: Pos,
    pub end: Pos,
    pub meta: &'a Meta<'a>,
}

impl PartialEq for Location<'_> {
    fn eq(&self, other: &Self) -> bool {
        self.start == other.start && self.end == other.end && self.meta.name == other.meta.name
    }
}

#[derive(Clone, Copy, PartialEq)]
pub struct Pos {
    pub line: i32,
    pub symbol: i32,
}

fn meta<'a>(name: &'a str, code: &'a str) -> Meta<'a> {
    Meta {
        name,
        lines: code.lines().collect(),
    }
}

File: src/Source.zig
const std = @import("std");

const Source = @This();

code: []const u8,
name: []const u8,
lines: []const []const u8,

pub fn read(io: std.Io, gpa: std.mem.Allocator, path: []const u8) !Source {
    const code = std.Io.Dir.cwd().readFileAlloc(io, path, gpa, .unlimited) catch {
        std.log.err("failed to read `{s}`", .{path});
        return error.Handled;
    };
    var vec = std.ArrayList([]const u8).empty;
    var iter = std.mem.splitAny(u8, code, "\r\n");
    while (iter.next()) |line| {
        try vec.append(gpa, line);
    }
    const lines = try vec.toOwnedSlice(gpa);
    return .{
        .code = code,
        .name = path,
        .lines = lines,
    };
}

pub fn deinit(source: Source, gpa: std.mem.Allocator) void {
    gpa.free(source.code);
    gpa.free(source.lines);
}
